pkg/cluster/update_controller.go

```diff
@@ -0,0 +1,290 @@
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	// ClusterConfigMapName is the name of the config map that holds cluster configuration,
+	// including the cluster version to run.
+	ClusterConfigMapName = "cluster-config"
+	// ClusterVersionKey is the key in the cluster ConfigMap that holds the version of the cluster
+	// that should be running.
+	ClusterVersionKey = "cluster.version"
+	// clusterMangedAnnotation is the annotation used to denote a managed component within
+	// the cluster.
+	clusterManagedLabel = "update-controller-managed=true"
+)
+
+type ComponentsGetterFn func(clientset.Interface, unversioned.Interface, cache.StoreToDaemonSetLister, cache.StoreToDeploymentLister, components.StoreToPodLister, cache.StoreToNodeLister) ([]Component, error)
+
+// UpdateController is responsible for safely updating an entire cluster.
+type UpdateController struct {
+	// Client is a generic API server client.
+	Client clientset.Interface
+	// UnversionedClient is another client, useful for events and
+	// not having to do conversions in certain places.
+	UnversionedClient unversioned.Interface
+	// AllNonNodeManagedComponentsFn is a function that should return
+	// a list of every non-Node component the update controller is managing.
+	GetAllManagedComponentsFn ComponentsGetterFn
+
+	// These stores hold all of the managed components.
+	nodes       cache.StoreToNodeLister
+	deployments cache.StoreToDeploymentLister
+	daemonSets  cache.StoreToDaemonSetLister
+
+	// pods managed by DaemonSets. Allows lookup by DaemonSet selector.
+	pods components.StoreToPodLister
+}
+
+// Component is responsible for updating
+// a single component in the cluster.
+// It takes the name of the component and a function
+// that will be used to update that component.
+// The name should be the name of the component
+// as it appears in the manifest file.
+type Component interface {
+	// Name is the name of the component to update.
+	Name() string
+	// UpdateToVersion is the function used to update this component to the
+	// provided version.
+	UpdateToVersion(*components.Version) (bool, error)
+	// Priority is the priority level for this component.
+	Priority() int
+	// Version of the component.
+	Version() (*components.Version, error)
+}
+
+// NewClusterUpdater returns a ClusterUpdater struct with defaults.
+func NewClusterUpdater(client clientset.Interface, uc unversioned.Interface) (*UpdateController, error) {
+	l, err := labels.Parse(clusterManagedLabel)
+	if err != nil {
+		return nil, err
+	}
+	mlo := api.ListOptions{LabelSelector: l}
+	nodeStore, nodeController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(_ api.ListOptions) (runtime.Object, error) {
+				return uc.Nodes().List(mlo)
+			},
+			WatchFunc: func(_ api.ListOptions) (watch.Interface, error) {
+				return uc.Nodes().Watch(mlo)
+			},
+		},
+		&api.Node{},
+		30*time.Minute,
```

Comment:
We probably want a much lower resync period for all of these informers. Maybe every few minutes? Also can we add it as a const.